presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java

```diff
@@ -242,52 +187,42 @@ private void doAbort()
-        }
+        int[] writerIndexes = getWriterIndexes(page);
 
         Block[] dataBlocks = getDataBlocks(page, sampleWeightBlock);
-        if (!bucketCount.isPresent()) {
-            if (pageIndexer.getMaxIndex() >= writers.length) {
-                writers = Arrays.copyOf(writers, pageIndexer.getMaxIndex() + 1);
-            }
-            for (int position = 0; position < page.getPositionCount(); position++) {
-                int writerIndex = indexes[position];
-                HiveWriter writer = writers[writerIndex];
-                if (writer == null) {
-                    writer = writerFactory.createWriter(partitionColumns, position, OptionalInt.empty());
-                    writers[writerIndex] = writer;
-                }
+        for (int position = 0; position < page.getPositionCount(); position++) {
+            int writerIndex = writerIndexes[position];
+            HiveWriter writer = writers.get(writerIndex);
+            writer.addRow(dataBlocks, position);
+        }
 
-                writer.addRow(dataBlocks, position);
-            }
+        return NOT_BLOCKED;
+    }
+
+    private int[] getWriterIndexes(Page page)
+    {
+        Page partitionColumns = extractColumns(page, partitionColumnsInputIndex);
+        Block bucketBlock = buildBucketBlock(page);
+        int[] writerIndexes = pagePartitioner.partitionPage(partitionColumns, bucketBlock);
+        if (pagePartitioner.getMaxIndex() >= maxOpenWriters) {
+            throw new PrestoException(HIVE_TOO_MANY_OPEN_PARTITIONS, "Too many open partitions");
         }
-        else {
-            for (int i = bucketWriters.size(); i <= pageIndexer.getMaxIndex(); i++) {
-                bucketWriters.add(new Int2ObjectOpenHashMap<>());
-            }
 
-            Page bucketColumnsPage = extractColumns(page, bucketColumns);
-            for (int position = 0; position < page.getPositionCount(); position++) {
-                int writerIndex = indexes[position];
-                Int2ObjectMap<HiveWriter> writers = bucketWriters.get(writerIndex);
-                int bucket = bucketFunction.getBucket(bucketColumnsPage, position);
-                HiveWriter writer = writers.get(bucket);
-                if (writer == null) {
-                    if (bucketWriterCount >= maxOpenPartitions) {
-                        throw new PrestoException(HIVE_TOO_MANY_OPEN_PARTITIONS, "Too many open writers for partitions and buckets");
-                    }
-                    bucketWriterCount++;
-                    writer = writerFactory.createWriter(partitionColumns, position, OptionalInt.of(bucket));
-                    writers.put(bucket, writer);
-                }
+        // create missing writers
+        while (writers.size() <= pagePartitioner.getMaxIndex()) {
+            int newWriterIndex = writers.size();
+            int position = Ints.indexOf(writerIndexes, newWriterIndex);
```

Comment:
`page` can potentially have many rows. Let's not write O(N^2) code. (N is the number of rows. Every row is a unique partition.) Given the low maxOpenWriters right now, it probably won't cause a problem. But this feels like a land mine.

I would do:
```
  for (int position = 0; position < page.getPositionCount(); position++) {
    if (writers.size() > pagePartitioner.getMaxIndex()) {
      break;
    }
    if (writerIndexes[i] < writers.size()) {
      continue;
    }
    verify(position == writers.size());
    // create the writer
    writers.add(writer);
  }
  verify(writers.size() == pagePartitioner.getMaxIndex() + 1;
```